cpp/src/reductions/scan/ewm.cu

```diff
@@ -0,0 +1,374 @@
+                      pairs.end(),
+                      output.begin(),
+                      [=] __device__(pair_type<T> pair) -> T { return pair.second; });
+    auto itr = thrust::make_counting_iterator<size_type>(0);
+
+    thrust::transform_inclusive_scan(rmm::exec_policy(stream),
+                                     itr,
+                                     itr + input.size(),
+                                     pairs.begin(),
+                                     ewma_adjust_functor<T, false, false>{beta},
+                                     recurrence_functor<T>{});
+  }
+
+
+  thrust::transform(
+    rmm::exec_policy(stream),
+    pairs.begin(),
+    pairs.end(),
+    output.begin(),
+    output.begin(),
+    [] __device__(pair_type<T> pair, T numerator) -> T { return numerator / pair.second; });
+
+  return output;
+}
+
+template <typename T>
+rmm::device_uvector<T> compute_ewma_noadjust(column_view const& input,
+                                             T const beta,
+                                             rmm::cuda_stream_view stream,
+                                             rmm::device_async_resource_ref mr)
+{
+  rmm::device_uvector<T> output(input.size(), stream);
+  rmm::device_uvector<pair_type<T>> pairs(input.size(), stream);
+  rmm::device_uvector<cudf::size_type> nullcnt =
+    [&input, stream]() -> rmm::device_uvector<cudf::size_type> {
+    if (input.has_nulls()) {
+      return null_roll_up(input, stream);
+    } else {
+      return rmm::device_uvector<cudf::size_type>(input.size(), stream);
+    }
+  }();
+  // denominators are all 1 and do not need to be computed
+  // pairs are all (beta, 1-beta x_i) except for the first one
+
+  if (!input.has_nulls()) {
+    auto data = thrust::make_zip_iterator(
+      thrust::make_tuple(input.begin<T>(), thrust::make_counting_iterator<size_type>(0)));
+    thrust::transform_inclusive_scan(rmm::exec_policy(stream),
+                                     data,
+                                     data + input.size(),
+                                     pairs.begin(),
+                                     ewma_noadjust_functor<T, false>{beta},
+                                     recurrence_functor<T>{});
+
+  } else {
+    /*
+    In this case, a denominator actually has to be computed. The formula is
+    y_{i+1} = (1 - alpha)x_{i-1} + alpha x_i, but really there is a "denominator"
+    which is the sum of the weights: alpha + (1 - alpha) == 1. If a null is
+    encountered, that means that the "previous" value is downweighted by a
+    factor (for each missing value). For example this would y_2 be for one null:
+    data = {x_0, NULL, x_1},
+    y_2 = (1 - alpha)**2 x_0 + alpha * x_2 / (alpha + (1-alpha)**2)
+
+    As such, the pairs must be updated before summing like the adjusted case to
+    properly downweight the previous values. But now but we also need to compute
+    the normalization factors and divide the results into them at the end.
+    */
```

Comment:
This comment doesn't seem to be referring to the code below it. Perhaps it should be moved to one of the functors above?